File: sdks/python/src/integrations/github.py
"""
GitHub Integration for the zv1 engine.

Provides issue management via the GitHub REST API.
"""

from __future__ import annotations

import time
from typing import Any

from src.utilities.sanitize_api_call import emit_api_call_event

try:
    import httpx
    HAS_HTTPX = True
except ImportError:
    HAS_HTTPX = False


class GithubIntegration:
    def __init__(self, api_key: str, base_url: str = "https://api.github.com", timeout: float = 30.0) -> None:
        if not HAS_HTTPX:
            raise ImportError("httpx package is required for GitHub integration.")
        self.api_key = api_key
        self.base_url = base_url
        self.client = httpx.AsyncClient(
            timeout=timeout,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Accept": "application/vnd.github+json",
                "X-GitHub-Api-Version": "2022-11-28",
                "Content-Type": "application/json",
            },
        )

    async def _request(self, method: str, url: str, params: dict | None = None, json_data: dict | None = None) -> Any:
        start_time = int(time.time() * 1000)
        request_headers = dict(self.client.headers)
        try:
            response = await self.client.request(method, url, params=params, json=json_data)
            await emit_api_call_event(getattr(self, "_engine_config", None), {
                "timestamp": start_time, "integration": "github", "nodeId": None, "nodeType": None,
                "request": {"method": method, "url": url, "headers": request_headers, "body": json_data},
                "response": {"status": response.status_code, "statusText": response.reason_phrase or ""},
                "duration": int(time.time() * 1000) - start_time, "error": None,
            })
            if response.status_code >= 400:
                msg = response.json().get("message", response.reason_phrase) if response.content else response.reason_phrase
                raise Exception(f"GitHub API error ({response.status_code}): {msg}")
            return response.json()
        except Exception as e:
            if "GitHub API error" not in str(e):
                await emit_api_call_event(getattr(self, "_engine_config", None), {
                    "timestamp": start_time, "integration": "github", "nodeId": None, "nodeType": None,
                    "request": {"method": method, "url": url, "headers": request_headers, "body": json_data},
                    "response": {"status": 0, "statusText": "Error"},
                    "duration": int(time.time() * 1000) - start_time, "error": str(e),
                })
            raise

    async def create_issue(self, owner: str, repo: str, title: str, body: str | None = None,
                           labels: list[str] | None = None, assignees: list[str] | None = None) -> dict[str, Any]:
        data: dict[str, Any] = {"title": title}
        if body:
            data["body"] = body
        if labels:
            data["labels"] = labels
        if assignees:
            data["assignees"] = assignees
        return await self._request("POST", f"{self.base_url}/repos/{owner}/{repo}/issues", json_data=data)

    async def list_issues(self, owner: str, repo: str, state: str = "open", labels: str | None = None,
                          per_page: int = 30, page: int = 1) -> list[dict[str, Any]]:
        params: dict[str, Any] = {"state": state, "per_page": per_page, "page": page}
        if labels:
            params["labels"] = labels
        return await self._request("GET", f"{self.base_url}/repos/{owner}/{repo}/issues", params=params)

    async def close(self) -> None:
        await self.client.aclose()

File: sdks/nodejs/src/integrations/github.js
import axios from 'axios';
import { emitAPICallEvent } from '../utilities/sanitizeAPICall.js';

export default class GithubIntegration {
    constructor(apiKey, options = {}) {
        this.apiKey = apiKey;
        this.options = {
            baseURL: 'https://api.github.com',
            timeout: 30000,
            ...options
        };
    }

    async _request(config) {
        const requestHeaders = {
            'Authorization': `Bearer ${this.apiKey}`,
            'Accept': 'application/vnd.github+json',
            'X-GitHub-Api-Version': '2022-11-28',
            'Content-Type': 'application/json',
            ...config.headers
        };
        const startTime = Date.now();

        try {
            const response = await axios({
                ...config,
                headers: requestHeaders,
                timeout: this.options.timeout
            });

            await emitAPICallEvent(this._engineConfig, {
                timestamp: startTime, integration: 'github', nodeId: null, nodeType: null,
                request: { method: config.method, url: config.url, headers: requestHeaders, body: config.data || null },
                response: { status: response.status, statusText: response.statusText },
                duration: Date.now() - startTime, error: null
            });

            return response.data;
        } catch (error) {
            await emitAPICallEvent(this._engineConfig, {
                timestamp: startTime, integration: 'github', nodeId: null, nodeType: null,
                request: { method: config.method, url: config.url, headers: requestHeaders, body: config.data || null },
                response: { status: error.response?.status || 0, statusText: error.response?.statusText || 'Error' },
                duration: Date.now() - startTime, error: error.message
            });
            if (error.response) {
                const msg = error.response.data?.message || error.response.statusText;
                throw new Error(`GitHub API error (${error.response.status}): ${msg}`);
            }
            throw new Error(`GitHub API error: ${error.message}`);
        }
    }

    async createIssue(owner, repo, title, options = {}) {
        const data = { title };
        if (options.body) data.body = options.body;
        if (options.labels) data.labels = options.labels;
        if (options.assignees) data.assignees = options.assignees;

        return await this._request({
            method: 'POST',
            url: `${this.options.baseURL}/repos/${owner}/${repo}/issues`,
            data
        });
    }

    async listIssues(owner, repo, options = {}) {
        const params = {};
        if (options.state) params.state = options.state;
        if (options.labels) params.labels = options.labels;
        if (options.per_page) params.per_page = options.per_page;
        if (options.page) params.page = options.page;

        return await this._request({
            method: 'GET',
            url: `${this.options.baseURL}/repos/${owner}/${repo}/issues`,
            params
        });
    }
}
